package.json

```diff
@@ -13,21 +13,23 @@
             "properties": {
                 "execMap": {
                     "type": "object",
+                    "markdownDescription": "Set the executor by [language id](https://code.visualstudio.com/docs/languages/identifiers)",
                     "additionalProperties": {
                         "type": "string"
                     },
                     "default": {
-                        "js": "tsm $path"
+                        "javascript": "node $path",
+                        "typescript": "tsx $path"
```

Comment:
Okay, I'm totally good with this change (also have been using `tsx` instead of `tsm` recently). But I wonder what if user already has changed the setting to something like this:

```json
"js": "tsx $path"
```

Maybe we should preserve this support as legacy option?

Reply:
For now, the user count is not too many, we may not need to consider the migrate problems. Or you can mark extension version to 1.0.0 and add break changes to changelog.md

Reply:
As I see from my experience almost no one reads changelog, until something is broken :(
However the process of generating changelog and new version is automated, so I'll be sure to include this information as breaking change to ensure it gets bumped to 0.2.0 (next breaking semver)

Reply:
Maybe I'm different, I like read changelog to know what's the new feature. For npm packages, I will always read changelog when I upgrade version, just in case there are any issues that may arise.

Reply:
This is okay for npm packages, but vscode extensions don't follow semver and 99.9% users don't turn off auto-updates (its enabled out of the box).

I'm also thinking wether its even okay to change default here. While its okay for me and you there is a potential possibility that user had something dangerous installed as `tsx` executable, so I think that from this release it would better to not have default setting at all.

Yes, I see that original code runner also uses default in `code-runner.executorMap` but you need to understand that once this setting was touched by user (for example to change executor of javascript only):
1. Huge number of commands gets inlined in settings.json (not a case with this extenion though)
2. All default executors get frozen, which means that futher potential improvements (like this, tsm -> tsx) are also useless (even if user never had to use typescript before)

Instead, it would be big improvement for newcomes to suggest set executable, when its missing for specific language. By doing this:

1. User would never end up with executing command he wasn't aware of.
2. User would see what command is recommended for specific language, after what he might immediately change it and/or install the executable itself
3. Futher improvements to recommended executables won't break anything for existing users and would allow to use "better defaults" for new ones

So I see it in the following way:

```ts
const recommendedExecutables = {
    javascript: "node ${path}",
    typescript: "tsx ${path}"
}

const getExecByLanguageId = (languageId) => {
    const execMap = getExtensionSetting('execMap')
    const exec = execMap[languageId]
    if (exec === undefined) {
        const recommended = recommendedExecutables[languageId]
        if (!recommended) return
        const choice = await showWarningMessage(`Language ${languageId} is missing an executable, but there is a recommended one: ${recommended}`, 'Set and execute', 'Just set to settings.json')
        // write to settings.json
    }

    return exec
}
```

This is also why this extension doesn't have a default keybinding.

Another thing that I want to ask is too preserve `$path` support, it won't be hard for us to just replace `$path -> ${path}`, but unfortunately most users are dunno and could have troubles with this migration if he has a huge number of executables set up, so we need to provide a way for smoother migration for such users, since without this 100% breakage of this extension for existing users is guaranteed.

Reply:
I think that support recommend will make this extension more complex. I just remove the settings default and add more docs. Check latest commit. You can just implemets by yourself